clients/unshaded/src/test/java/tachyon/client/file/FileInStreamTest.java

```diff
@@ -0,0 +1,335 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.client.file;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.List;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import com.google.common.collect.Lists;
+
+import tachyon.Constants;
+import tachyon.client.ClientContext;
+import tachyon.client.TachyonStorageType;
+import tachyon.client.block.BlockInStream;
+import tachyon.client.block.BufferedBlockInStream;
+import tachyon.client.block.TachyonBlockStore;
+import tachyon.client.block.TestBufferedBlockInStream;
+import tachyon.client.block.TestBufferedBlockOutStream;
+import tachyon.client.file.options.InStreamOptions;
+import tachyon.client.util.ClientMockUtils;
+import tachyon.conf.TachyonConf;
+import tachyon.exception.ExceptionMessage;
+import tachyon.exception.PreconditionMessage;
+import tachyon.test.Tester;
+import tachyon.thrift.FileInfo;
+import tachyon.underfs.UnderFileSystem;
+import tachyon.util.io.BufferUtils;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({FileSystemContext.class, TachyonBlockStore.class, UnderFileSystem.class})
+public class FileInStreamTest implements Tester<FileInStream> {
+
+  private static final long BLOCK_LENGTH = 100L;
+  private static final long FILE_LENGTH = 350L;
+  private static final long NUM_STREAMS = ((FILE_LENGTH - 1) / BLOCK_LENGTH) + 1;
+
+  private TachyonBlockStore mBlockStore;
+  private TachyonConf mConf;
+  private FileSystemContext mContext;
+  private FileInfo mInfo;
+
+  private List<BufferedBlockInStream> mInStreams;
+  private List<TestBufferedBlockOutStream> mCacheStreams;
+
+  private FileInStream.PrivateAccess mPrivateAccess;
+  private FileInStream mTestStream;
+
+  @Override
+  public void receiveAccess(Object access) {
+    mPrivateAccess = (FileInStream.PrivateAccess) access;
+  }
+
+  @Rule
+  public final ExpectedException mThrown = ExpectedException.none();
+
+  @Before
+  public void before() throws IOException {
+    mInfo = new FileInfo().setBlockSizeBytes(BLOCK_LENGTH).setLength(FILE_LENGTH);
+
+    // Set small buffer sizes so that we don't run out of heap space
+    mConf = new TachyonConf();
+    mConf.set(Constants.USER_BLOCK_REMOTE_READ_BUFFER_SIZE_BYTES, "4KB");
+    mConf.set(Constants.USER_FILE_BUFFER_BYTES, "4KB");
+    ClientContext.reset(mConf);
```

Comment:
Is there a place we can consolidate this logic? It seems like it would be used in multiple places?